src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -65,12 +65,47 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
+
+            if (firstStepResult == null)
+            {
+                Authenticate(connection, description, previousStep: null, previousStepResult: null, cancellationToken);
+            }
+            else
+            {
+                var (firstStep, _) = CreateFirstStepAndCommand(_mechanism, connection, conversation: null, description);
+                Authenticate(
+                    connection,
+                    description,
+                    previousStep: firstStep,
+                    previousStepResult: firstStepResult,
+                    cancellationToken);
+            }
+        }
+
+        private void Authenticate(
+            IConnection connection,
+            ConnectionDescription description,
+            ISaslStep previousStep,
+            BsonDocument previousStepResult,
+            CancellationToken cancellationToken)
+        {
+            Ensure.IsNotNull(connection, nameof(connection));
+            Ensure.IsNotNull(description, nameof(description));
+            if (previousStep != null)
+            {
+                Ensure.IsNotNull(previousStepResult, nameof(previousStepResult));
+            }
+
             using (var conversation = new SaslConversation(description.ConnectionId))
             {
-                var currentStep = _mechanism.Initialize( connection, conversation, description);
+                /* We cannot skip the cast to StepAndCommand below without adding a reference to ValueTuple */
```

Comment:
We're able to use this tiny bit of syntactic sugar thanks to the `Deconstruct` method here: https://github.com/vincentkam/mongo-csharp-driver/pull/95/files#diff-7cb6644a6c07129601487f26bb0f9b42R477

For some reason, skipping the cast results in this error :
```
Core\Authentication\SaslAuthenticator.cs(104,23): error CS8128: Member '.ctor' was not found on type '(, )' from assembly 'MongoDB.Driver.Core, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
[C:\Users\Vincent\driver\src\MongoDB.Driver.Core\MongoDB.Driver.Core.csproj]
Core\Authentication\SaslAuthenticator.cs(105,23): error CS8128: Member '.ctor' was not found on type '(, )' from assembly 'MongoDB.Driver.Core, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
[C:\Users\Vincent\driver\src\MongoDB.Driver.Core\MongoDB.Driver.Core.csproj]
Core\Authentication\SaslAuthenticator.cs(104,23): error CS8128: Member '.ctor' was not found on type '(, )' from assembly 'MongoDB.Driver.Core, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
[C:\Users\Vincent\driver\src\MongoDB.Driver.Core\MongoDB.Driver.Core.csproj]
Core\Authentication\SaslAuthenticator.cs(105,23): error CS8128: Member '.ctor' was not found on type '(, )' from assembly 'MongoDB.Driver.Core, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.
[C:\Users\Vincent\driver\src\MongoDB.Driver.Core\MongoDB.Driver.Core.csproj]```
```
This was confusing to me, so I added the comment to clarify why this happens.

Previously, I got around this error by splitting the line into two statements: https://github.com/vincentkam/mongo-csharp-driver/pull/95#discussion_r401101326

Reply:
Oudated with the refactoring that combines everyone's approaches :)